oceanstream/L3_regridded_data/shoal_process.py

```diff
@@ -0,0 +1,190 @@
+import scipy.ndimage as nd_img
+import xarray as xr
+from echopype.mask.api import apply_mask
+from haversine import haversine
+from pandas import DataFrame
+
+
+def split_shoal_mask(Sv: xr.Dataset):
+    """
+    Given a Sv dataset with an existing shoal mask, generates a list of
+    individual masks for each shoal
+
+    Parameters:
+    - Sv: xr.Dataset - Sv dataset with an existing shoal mask
+
+    Returns:
+    - [xr.DataArray]: list of individual shoal masks
+
+    Example:
+        >>> split_shoal_mask(Sv)
+    """
+    all_shoals = Sv["mask_shoal"]
+    labelled = nd_img.label(all_shoals)
+    la = labelled[0]
+    ls = labelled[1]
+    shoals = []
+    for i in range(1, ls + 1):
+        shoal = xr.DataArray(
+            data=(la == i),
+            dims=all_shoals.dims,
+            coords=all_shoals.coords,
+        )
+        shoal.attrs["label"] = i
+        shoals.append(shoal)
+    return shoals
+
+
+def process_single_shoal(Sv: xr.Dataset, mask: xr.DataArray):
+    """
+    Given a Sv dataset and a shoal mask containing a single shoal, returns a
+    dataframe containing shoal metadata for each channel
+
+    Parameters:
+    - Sv: xr.Dataset - Sv dataset
+    - mask: xr.DataArray - single-shoal mask
+
+    Returns:
+    - [pd.Dataframe]: a list of single-row-per-channel dictionaries containing shoal metadata
+
+    Example:
+        >>> process_single_shoal(Sv, mask)
+    """
+    channels = Sv["channel"]
+    results = [process_single_shoal_channel(Sv, mask, c) for c in channels]
+    results = [r for r in results if r is not None]
+    if results == []:
+        return None
+    return results
+
+
+def process_single_shoal_channel(Sv: xr.Dataset, mask: xr.DataArray, channel: str):
+    """
+    Given a Sv dataset, a shoal mask containing a single shoal and the desired channel,
+    returns a dataframe containing shoal metadata for that specific channel
+
+    Parameters:
+    - Sv: xr.Dataset - Sv dataset
+    - mask: xr.DataArray - single-shoal mask
+    - channel: str - channel
+
+    Returns:
+    - dict: a dictionary containing shoal metadata
+
+    Example:
+        >>> process_single_shoal_channel(Sv, mask, channel)
+    """
+    mc = mask.sel(channel=channel)
+    Sv_sel = Sv.sel(channel=channel)
+    Sv_masked = apply_mask(Sv_sel.copy(deep=True), mc)
+
+    # create a dataset that has the mask instead of Sv
+    md = Sv_masked.copy(deep=True)
+    md["Sv"] = ~md["Sv"].isnull()
+
+    Sv_mean = Sv_masked.Sv.mean(skipna=True).values
+    frequency = Sv_masked.frequency_nominal.values
+    area = mc.sum().values
+    if area == 0:
+        return None
+    filename = md.source_filenames.values
+    label = mc.attrs["label"]
+
+    ping_true = md["Sv"].any(dim="range_sample")
+    range_true = md["Sv"].any(dim="ping_time")
+    subset_md = md.sel(ping_time=ping_true, range_sample=range_true)
+
+    # TODO figure out exactly how the shiny tool does plotting
+    # bbox_0 = subset_md.ping_time.min().values
```

Comment:
I think the bounding boxes should be:
```py
bbox_0 = float(subset_md.ping_time.min())
bbox_1 = float(subset_md.range_sample.min())
bbox_2 = float(subset_md.ping_time.max())
bbox_3 = float(subset_md.range_sample.max())
```

Reply:
So did I, at first, but this is not what the toolset consuming these is using. This portion can be ignored for now, I'll fix it in a further commit today